sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IO.java

```diff
@@ -0,0 +1,658 @@
+              .setCoder(new HealthcareIOErrorCoder<>(new HL7v2MessageCoder()));
+      return Write.Result.in(input.getPipeline(), failedInserts);
+    }
+
+    /** The type Write hl 7 v 2 fn. */
+    static class WriteHL7v2Fn extends DoFn<HL7v2Message, HealthcareIOError<HL7v2Message>> {
+      // TODO when the healthcare API releases a bulk import method this should use that to improve
+      // throughput.
+
+      private Counter failedMessageWrites =
+          Metrics.counter(WriteHL7v2Fn.class, "failed-hl7v2-message-writes");
+      private final String hl7v2Store;
+      private final Counter throttledSeconds =
+          Metrics.counter(WriteHL7v2Fn.class, "cumulative-throttling-seconds");
+      private final Counter successfulHL7v2MessageWrites =
+          Metrics.counter(WriteHL7v2.class, "successful-hl7v2-message-writes");
+      private transient AdaptiveThrottler throttler;
+      private final Write.WriteMethod writeMethod;
+
+      private static final Logger LOG = LoggerFactory.getLogger(WriteHL7v2.WriteHL7v2Fn.class);
+      private transient HealthcareApiClient client;
+
+      /**
+       * Instantiates a new Write HL7v2 fn.
+       *
+       * @param hl7v2Store the HL7v2 store
+       * @param writeMethod the write method
+       */
+      WriteHL7v2Fn(String hl7v2Store, Write.WriteMethod writeMethod) {
+        this.hl7v2Store = hl7v2Store;
+        this.writeMethod = writeMethod;
+      }
+
+      /**
+       * Init client.
+       *
+       * @throws IOException the io exception
+       */
+      @Setup
+      public void initClient() throws IOException {
+        this.client = new HttpHealthcareApiClient();
+      }
+
+      @StartBundle
+      public void startBundle() {
+        if (throttler == null) {
+          throttler = new AdaptiveThrottler(1200000, 10000, 1.25);
+        }
+      }
+
+      /**
+       * Write messages.
+       *
+       * @param context the context
+       */
+      @ProcessElement
+      public void writeMessages(ProcessContext context) {
+        HL7v2Message msg = context.element();
+        final int throttleWaitSeconds = 5;
+        long startTime = System.currentTimeMillis();
+        Sleeper sleeper = Sleeper.DEFAULT;
+        switch (writeMethod) {
+          case BATCH_IMPORT:
+            throw new UnsupportedOperationException("The Batch import API is not available yet");
+          case INGEST:
+          default:
+            try {
+              if (throttler.throttleRequest(startTime)) {
+                LOG.info("Delaying request due to previous failures.");
+                this.throttledSeconds.inc(throttleWaitSeconds);
+                sleeper.sleep(throttleWaitSeconds * 1000);
+                this.throttler.successfulRequest(startTime);
+                this.successfulHL7v2MessageWrites.inc();
+              }
+              client.ingestHL7v2Message(hl7v2Store, msg.toModel());
+            } catch (Exception e) {
+              failedMessageWrites.inc();
```

Comment:
I believe the reason for failure may be relevant.  a 429 would want a certain number of retries, where a 401 or 403 may need to refresh the credentials, where another error is likely a hard stop.

Reply:
all retryable errors will be retried by the request initializer in the http client.

I think adding the throttling was a premature optimization on my part. We should remove adaptive throttling entirely and test this at load and see if throttling is necessary.